sdk/core/core/inc/az_precondition_failed.h

```diff
@@ -39,7 +39,7 @@ AZ_NODISCARD AZ_INLINE bool az_span_is_valid(az_span span, int32_t min_length)
      or span length is greater than 0 and pointer is not NULL, and greater or equal to min_length
   */
   return (
-      (span_length == 0
+      ((az_span_ptr(span) != NULL && az_span_length(span) == 0 && az_span_capacity(span) > 0)
```

Comment:
this would means ptr can never be NULL right?

what about
```c
az_span body = { 0 };
// if az_span ptr can't be NULL, then az_http_response_get_body can't be set for valid spans only
AZ_RETURN_IF_FAILED(az_http_response_get_body(&response, &body));
```

Reply:
No it can be `NULL` in the case of an `AZ_SPAN_NULL` since in that case the capacity is 0.

Reply:
This would also impact the way az_span_copy works
```c
/**
* We currently support trying to copy src span to a dst span with 0 capacity.
* If we consider 0 capacity and 0 length as non valid, instead of returning OVERFLOW, we would be sleeping forever
*/
AZ_NODISCARD az_result az_span_copy(az_span dst, az_span src, az_span* out)
{
  AZ_PRECONDITION_VALID_SPAN(dst, 0);
  AZ_PRECONDITION_VALID_SPAN(src, 0);
```

Reply:
Here's the logic I recommend; it is very explicit:
// A span is valid if:
ptr == NULL && len == 0 && cap == 0	// AZ_SPAN_NULL
// OR 
ptr != NULL && len >=0 && cap >= 0 && len <= cap	// not AZ_SPAN_NULL

Reply:
I almost agree. If the pointer is not `NULL` then shouldn't the capacity be more than 0?

Reply:
No, the capacity can still be 0. Logically, it is legal and it can happen if you slice of the remainder (right edge) of a buffer and you reached the end of the buffer

Reply:
If I understand the code correctly that isn't what currently happens though. 
https://github.com/Azure/azure-sdk-for-c/blob/3094430d27c9bbf5461082d5070f3c91969822c6/sdk/core/core/src/az_span_private.h#L40-L45
In this case if you drop the total capacity you get an `AZ_SPAN_NULL`

Reply:
we have get rid of az_span_drop and az_span_take in the CONTRACT to PRECONDITION PR.

Precondition allows `az_span_slice()` to return az_span directly instead of az_result and be considered as never failed operation

so doing something like
```c
az_span some_span = AZ_SPAN_FROM_STR("1");
az_span next_position = az_span_slice(some_span, az_span_length(some_span) + 1, -1);
```
This could return an span with length and capacity 0 if `some_span` was length and capacity = 1.

Reply:
Should that be the behavior though? Because the returned span (with a valid address but length and capacity of 0) is seeming to have a weird property. It has a "valid" address but with nothing there. Why not just return an `AZ_SPAN_NULL`.
The C99 spec also doesn't allow for 0 length arrays which is roughly what this is doing.
***C99 6.7.5.2***
> If they delimit an expression (which specifies the size of an array), the
expression shall have an integer type. If the expression is a constant expression, it shall
have a value greater than zero.

Reply:
it is indeed weird to have a valid address but len and cap equal to 0.

The only advantage I can see, would be if we want to somehow go back, we would still have the ptr. But this can also happen if we don't mutate previous span..
Example
```c
az_span mutating = AZ_SPAN_FROM_STR("1");
mutating = az_span_slice(mutating, az_span_length(mutating) + 1, -1);
// at this point, we could still go back if we want
// But if we'd have returned AZ_SPAN_NULL here, we would have lost the address
if (az_span_capacity(mutating) == 0) {
  mutating = az_span_init(az_span_ptr(mutating) -1, 1, 1);
}

// Yes, this can be avoided by non-mutating original az_span
```
 
Let's define this in Friday meeting as well.  What is it going to be a valid span

Reply:
But you can't go back, because even if you have the pointer, there's no information in a `span` which tells you data previous to the pointer is valid for use.